[Streams] Update Streams IA and structure. (#7062)
elastic/docs-content: solutions/observability/streams-new/parse-and-process.md

---
navigation_title: Process documents
applies_to:
  serverless: ga
  stack: preview =9.1, ga 9.2+
description: Process and extract fields from incoming Streams documents using configurable processors, conditions, and live simulation previews.
products:
  - id: observability
  - id: elasticsearch
  - id: kibana
  - id: cloud-serverless
  - id: cloud-hosted
  - id: cloud-enterprise
  - id: cloud-kubernetes
  - id: elastic-stack
---

# Process your documents with Streams [streams-extract-fields]

Most log data arrives as unstructured text. To filter, search, and analyze it effectively, you need to extract fields from that raw content. For example, extracted fields let you filter for log messages with an `ERROR` log level that occurred during a specific time period to help diagnose an issue.

> Streams **Processing** tab provides a single place to build and manage your document processing pipeline:

Reviewer: ```suggestion
The Streams **Processing** tab provides a single place to build and manage your document processing pipeline:
```

- **[Add processors and conditions](#streams-add-processors)**: Use the Streams UI without needing to manually configure pipeline JSON or Grok syntax.

> - {applies_to}`serverless: preview` {applies_to}`stack: preview 9.3+` **[Generate pipeline suggestions using AI](#streams-generate-pipeline-suggestions)**: Let Streams analyze sample documents and suggests pipeline patterns, so you're refining instead of writing from scratch.

Reviewer: ```suggestion
- {applies_to}`serverless: preview` {applies_to}`stack: preview 9.3+` **[Generate pipeline suggestions using AI](#streams-generate-pipeline-suggestions)**: Let Streams analyze sample documents and suggest pipeline patterns, so you're refining instead of writing from scratch.
```

> - {applies_to}`serverless: preview` {applies_to}`stack: preview 9.3+` **[Generate pipeline suggestions using AI](#streams-generate-pipeline-suggestions)**: Let Streams analyze sample documents and suggests pipeline patterns, so you're refining instead of writing from scratch.

Reviewer 2: Can we blend these 2 entries together? Maybe something along the lines of:

```
Add processors and conditions with pipeline suggestions: Use the Streams UI without needing to manually configure pipeline JSON or Grok syntax, and let Streams analyze sample documents to suggest pipeline patterns, so you're refining instead of writing from scratch.
```

(feel free to reword, it's not ideal)

- **[Preview changes](#streams-preview-changes)**: Use the data preview to view which fields your pattern extracts per document, so you can verify field extraction before saving.
- **[Detect and resolve processing issues](#streams-detect-failures)**: Identify which processor or condition is causing documents to fail during processing.
- **[Catch mapping conflicts](#streams-detect-failures)**: Identify potential mapping conflicts before they cause cluster-wide failures. Streams simulates the indexing process end-to-end before deploying.

## Add and configure processors [streams-add-processors]

> Streams uses [{{es}} ingest pipelines](../../../manage-data/ingest/transform-enrich/ingest-pipelines.md) made up of processors and conditions to transform your data, without requiring you to switch interfaces and manually update pipelines.

Reviewer 2: Can we say "underneath"? As the language that is primarily used isn't actually ingest pipeline, but an abstracted layer called Streamlang (hence the slight difference in syntax and capabilities).

:::::::{stepper}
::::::{step} Open the Processing tab
1. Open **Streams** from the navigation menu or use the [global search field](../../../explore-analyze/find-and-organize/find-apps-and-objects.md).
1. Select your stream from the list.
1. Go to the **Processing** tab.
::::::

::::::{step} Add processors and conditions
:anchor: streams-generate-pipeline-suggestions

Use any combination of the following options to build your processing pipeline:

- **Suggest a pipeline**: Let Streams analyze your sample data and generate a complete processor pipeline using AI.
- **Manually add processors**: Select and configure individual processors yourself when you know which transformations you need.
- **Add conditions**: Attach Boolean expressions to define when to run processors.

::::{dropdown} Suggest a pipeline

```{applies_to}
stack: preview 9.3+
serverless: preview
```

:::{note}
This feature requires a [Generative AI connector](kibana://reference/connectors-kibana/gen-ai-connectors.md).
:::

> Setting up processors is generally a multi-step process. For example, you might need a grok processor to extract fields, a date processor to convert timestamps, and a remove processor to remove temporary fields. Instead of creating individual processors manually, you can have AI suggest an entire pipeline for you:

Reviewer: There is some inconsistency across the docs with regard to this being `Grok processor` or `grok processor`. Can we standardize the (lack of?) capitalization?

1. Review the suggested processors, and either **Accept** or **Reject** the suggestions.
1. Select **Regenerate** to have Streams regenerate the suggested pipeline. Change the large language model (LLM) that Streams uses to generate suggestions from the {icon}`controls` menu.

:::{dropdown} How does Suggest a pipeline work?
:::{include} ../../_snippets/streams-suggestions.md
:::
:::
::::

::::{dropdown} Manually add processors

If you know which processors you want to use, you can add them manually from the Streams UI. Refer to the [Streamlang reference](./streamlang.md) for supported processors and configuration details.

1. Select **Create processor**. To have Streams suggest processors instead, select [Suggest a pipeline](#streams-generate-pipeline-suggestions).
1. Select a processor from the **Processor** menu.

   :::{note}
   Let Streams suggest patterns for [Grok](./processors/grok.md#streams-grok-patterns) and [dissect](./processors/dissect.md#streams-dissect-patterns) processors by selecting **Generate pattern**. This feature requires a [Generative AI connector](kibana://reference/connectors-kibana/gen-ai-connectors.md).
   :::

> 1. Configure the processor and select **Create** to save the processor.

Reviewer: Maybe:

```suggestion
1. Configure the processor, then select **Create** to save it.
```

1. Optional: Turn on **Ignore failures** if you want document processing to continue even when this processor fails.
1. Optional: For dissect, Grok, and rename processors, turn on **Ignore missing fields** if you want processing to continue when a source field is missing.
::::

::::{dropdown} Add conditions

You can add Boolean expressions evaluated for each document and attach processors that run only when those conditions are met.

To add a condition:

1. Select **Create** → **Create condition**.
1. Provide a **Field**, a **Value**, and a comparator.
1. Select **Create condition**.

> 1. After creating a condition, add a processor or another condition to it by selecting the {icon}`plus_in_circle`.

Reviewer: maybe:

```suggestion
1. After creating a condition, add a processor or another condition to it by selecting the {icon}`plus_in_circle` icon.
```

Refer to [Conditions](./streamlang.md#streams-streamlang-conditions) in the Streamlang reference for supported operators and examples.
::::
::::::

::::::{step} Preview changes
:anchor: streams-preview-changes

After adding processors and conditions, the **Data preview** tab simulates processor results with additional filtering options depending on the outcome of the simulation.

> When you add or edit processors, the **Data preview** tab updates automatically.

Reviewer: Would it make more sense to switch the places of these two sentences?

```suggestion
When you add or edit processors, the **Data preview** tab updates automatically.

After adding processors and conditions, the **Data preview** tab simulates processor results with additional filtering options depending on the outcome of the simulation.
```

> The **Data preview** tab loads 100 documents from your existing data and runs your changes against them.

Reviewer 2: There's also a way to "select" a different sample of documents, can we add a quick blurb about this?

Reviewer 3: Are we talking about the "Refetch samples and rerun simulation" option here?

Reviewer 2: Not exactly, check this out:

https://github.com/user-attachments/assets/f2e5a7a2-c9fa-419c-b50c-5b4411364d7a

I think we can improve the UX, but it's besides the point, the capability is there ;)

To switch modes, select the appropriate tab from the top of the processing page.

:::{image} ../../images/streams-editing-modes.png
:screenshot:
:::

Streams defaults to interactive mode unless the configuration can't be represented in interactive mode (for example, when nesting levels are too deep).

### When to use Interactive mode [streams-editing-interactive-mode]

**Interactive** mode provides a form-based interface for creating and editing processors. This mode works best for:

- Users who prefer a guided, visual approach
- Configurations that don't require deeply nested conditions

### When to use YAML mode [streams-editing-yaml-mode]
```{applies_to}
stack: ga 9.3+
```

**YAML** mode provides a code editor for writing Streamlang directly. This mode works best for:

- Users who prefer working with code
- Advanced configurations with complex or deeply nested conditions

Refer to the [Streamlang reference](./streamlang.md) for the complete syntax, condition operators, and examples.

## Known limitations [streams-known-limitations]

> - Streams does not support all processors. Refer to the [Streamlang reference](./streamlang.md) for supported processors.

Reviewer: Does this refer to all ingest processors? If yes, I think it’s worth to add here:

```suggestion
- Streams does not support all ingest processors. Refer to the [Streamlang reference](./streamlang.md) for supported processors.
```

- The data preview simulation might not accurately reflect the changes to the existing data when editing existing processors or re-ordering them.
- Streams can't properly handle arrays. Although it supports basic actions like appending or renaming, it can't access individual array elements. For classic streams, the workaround is to use the [manual pipeline configuration](./processors/manual-pipeline-configuration.md) that supports {{product.painless}} scripting and all ingest processors.

## How Streams applies processing changes [streams-applied-changes]

When you save processors in the Streams UI, Streams automatically manages the underlying ingest pipeline configuration. You don't need to edit pipeline JSON directly. Streams identifies the best-matching pipeline for your data stream, appends your processing logic to it, and maintains a dedicated managed pipeline to keep your changes isolated and reproducible. Expand the following section for more detail on how Streams applies processing changes.

> ::::{dropdown} Automatic pipeline management

Reviewer 2: ```suggestion
::::{dropdown} Implementation details: automatic pipeline management
```

Or feel free to word it differently :)

Streams identifies the appropriate `@custom` pipeline (for example, `logs-myintegration@custom` or `logs@custom`) by checking the `default_pipeline` that is set on the data stream. You can view the default pipeline on the **Advanced** tab under **Ingest pipeline**.

In this default pipeline, Streams locates the last processor that calls a pipeline ending in `@custom`.
- For integrations, this would result in a pipeline name like `logs-myintegration@custom`.
- Without an integration, the only `@custom` pipeline available might be `logs@custom`.

If no default pipeline is detected, Streams adds a default pipeline to the data stream by updating the index templates.

If a default pipeline is detected, but it does not contain a custom pipeline, Streams adds the pipeline processor directly to the pipeline.

Streams then adds a pipeline processor to the end of that `@custom` pipeline. This processor definition directs matching documents to a dedicated pipeline managed by Streams called `<data_stream_name>@stream.processing`:

```json
// Example processor added to the relevant @custom pipeline
{
  "pipeline": {

>     "name": "<data_stream_name>@stream.processing", // for example, logs-my-app-default@stream.processing

Reviewer: Because this is rendered as a code annotation below the snippet, maybe:

```suggestion
    "name": "<data_stream_name>@stream.processing", // For example, `logs-my-app-default@stream.processing`
```
